src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs

```diff
@@ -20,35 +20,108 @@ internal sealed class ModelTypeProvider : TypeProvider
+        {
+            get
+            {
+                if (_publicConstructor != null)
+                {
+                    return _publicConstructor;
+                }
+
+                _fieldsInitializer();
+                return _publicConstructor!;
+            }
+        }
+        public ConstructorSignature SerializationConstructor
+        {
+            get
+            {
+                if (_serializationConstructor != null)
+                {
+                    return _serializationConstructor;
+                }
+
+                _fieldsInitializer();
+                return _serializationConstructor!;
+            }
+        }
+
+        private IReadOnlyDictionary<FieldDeclaration, InputModelProperty> FieldsToInputs
+        {
+            get
+            {
+                if (_fieldsToInputs != null)
+                {
+                    return _fieldsToInputs;
+                }
 
-        private readonly IReadOnlyDictionary<FieldDeclaration, InputModelProperty> _fieldsToInputs;
+                _fieldsInitializer();
+                return _fieldsToInputs!;
+            }
+        }
         // parameter name should be unique since it's bound to field property
-        private readonly IReadOnlyDictionary<string, FieldDeclaration> _parameterNamesToFields;
+        private IReadOnlyDictionary<string, FieldDeclaration> ParameterNamesToFields
+        {
+            get
+            {
+                if (_parameterNamesToFields != null)
+                {
+                    return _parameterNamesToFields;
+                }
+
+                _fieldsInitializer();
+                return _parameterNamesToFields!;
+            }
+        }
+
+        // lazy envaluation
+        private readonly Action _fieldsInitializer;
+        private IReadOnlyList<FieldDeclaration>? _fields;
+        private ConstructorSignature? _publicConstructor;
+        private ConstructorSignature? _serializationConstructor;
+        private IReadOnlyDictionary<FieldDeclaration, InputModelProperty>? _fieldsToInputs;
+        // parameter name should be unique since it's bound to field property
+        private IReadOnlyDictionary<string, FieldDeclaration>? _parameterNamesToFields;
 
         public ModelTypeProvider(InputModelType inputModel, TypeFactory typeFactory, string defaultNamespace, SourceInputModel? sourceInputModel)
             : base(inputModel.Namespace ?? defaultNamespace, sourceInputModel)
         {
             DefaultName = inputModel.Name;
             DefaultAccessibility = inputModel.Accessibility ?? "public";
 
-            (_fieldsToInputs, var serializationParameters, _parameterNamesToFields) = CreateParametersAndFieldsForRoundTripModel(inputModel, typeFactory);
+            _fieldsInitializer = delegate()
+            {
+                (_fieldsToInputs, var serializationParameters, _parameterNamesToFields) = CreateParametersAndFieldsForRoundTripModel(inputModel, typeFactory);
```

Comment:
It this also because we don't have access to field names upfront?

Reply:
> It this also because we don't have access to field names upfront?

Sort of. Each field of a `ModelTypeProvider` is a `TypeProvider`. We resolve them in https://github.com/Azure/autorest.csharp/blob/c3e7e60ef50975a5dbc21c9ddd3eb484522f1949/src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs#L84

The problem is that at that time the `_library` is not initialized. There are basically two options here:
- Use lazy evaluation, this is the approach I chose
- Need to traverse the type maps and get the dependency sequence. Then initialize from the types with least dependency first.

Reply:
I've added simple model support in 2559: https://github.com/Azure/autorest.csharp/pull/2559/files#diff-d8015b998fd7c30224b9fff88beb8b080bcb5f7552c821d861c16c35cbdc83c8R31

Can you please try in after 2559 is merged, so we don't have to use lazy initialization here?

Full "model with model properties" support isn't there yet.

Reply:
> I've added simple model support in 2559: [#2559 (files)](https://github.com/Azure/autorest.csharp/pull/2559/files#diff-d8015b998fd7c30224b9fff88beb8b080bcb5f7552c821d861c16c35cbdc83c8R31)
> 
> Can you please try in after 2559 is merged, so we don't have to use lazy initialization here?
> 
> Full "model with model properties" support isn't there yet.

Verified in https://github.com/Azure/autorest.csharp/pull/2572 Thanks.